src/metabase/api/search.clj

```diff
@@ -374,23 +303,43 @@
             (if (number? v)
               (not (zero? v))
               v))]
-    (let [search-query {:union-all (for [model searchable-models
+    (let [search-query {:union-all (for [model search-config/searchable-models
                                          :let  [query (search-query-for-model model search-ctx)]
                                          :when (seq query)]
                                      query)}
           _            (log/tracef "Searching with query:\n%s" (u/pprint-to-str search-query))
-          results      (db/query search-query :max-rows search-max-results)]
-      ;; sort by [score model name]
-      (sort-by (juxt
-                (comp - (partial search/score (:search-string search-ctx)))
-                (comp model->sort-position :model)
-                :name)
-               (for [row results
-                     :when (check-permissions-for-model row)]
-                 ;; MySQL returns `:favorite` and `:archived` as `1` or `0` so convert those to boolean as needed
-                 (-> row
-                     (update :favorite bit->boolean)
-                     (update :archived bit->boolean)))))))
+          results      (db/reducible-query search-query :max-rows search-config/db-max-results)
+          cmp          (reify java.util.Comparator
+                         (compare [_ [score-1 _result-1] [score-2 _result-2]]
+                           (compare score-1 score-2)))
```

Comment:
this will be problematic i think. i _think_ (not positive) that if it compares two things as equal then it considers them as equal, not just of equal rank. This might drop things. Need to be super careful about that.

Reply:
disregard. the queue can handle multiple equal things. we are golden here. might be a problem in the tree set but not in our priority queue.

Reply:
If two things are "equal" then compare will return `0`. Because of this:

```clojure
                          (if (>= (.size q) search-config/max-filtered-results)
                            (let [smallest (.peek q)]
                              (if (pos? (.compare cmp item smallest))
                                (doto q
                                  (.poll)
                                  (.offer item))
                                q))
```

that matters when a new item is equal to the smallest item in the full PQ, and since it's `pos?` instead of of non-negative we'll favor the item that's already in there.